packages/community-cli-plugin/src/utils/loadMetroConfig.js

```diff
@@ -71,6 +72,17 @@ function getOverrideConfig(
       ],
     },
   };
+
+  // Applying the heuristic of appending workspace root as watch folder,
+  // only if no other watch folder (beside the project root) has been given.
+  if (!config.watchFolders.some(folder => folder !== ctx.root)) {
```

Comment:
The only potential problem I see with this approach of providing the default as an override after loading the user-land config, is that the developer has to provide at least one element (other than the project root which is automatically injected) in the watchFolders to disable this default heuristic.

Reply:
Yeah it's tricky 😅.

I think let's bypass this logic if `config.watchFolders != null`. i.e. _**any value set by the user is an opt-out**_. We'd also want to note this [in the docs](https://reactnative.dev/docs/metro) — under a header "Workspaces detection" or similar (which could also briefly describe this feature to users 🙂).

Reply:
I did some experiments passing the workspace root via the second parameter to `metro-config`'s `loadConfig` function.

It turns out that even if we omitted `watchFolders` from the config here: https://github.com/facebook/react-native/blob/f1a7f08feb2a63926e91d37b5d85130c7b74ee31/packages/metro-config/index.js#L85 it would still propagate as `[]` to the `community-cli-plugin` because the `metro-config` package as `watchFolders: []` here: https://github.com/facebook/metro/blob/main/packages/metro-config/src/defaults/index.js#L163 and we merge on top of that:
https://github.com/facebook/react-native/blob/f1a7f08feb2a63926e91d37b5d85130c7b74ee31/packages/metro-config/index.js#L91-L94

I tried explicitly setting the `watchFolders` to `undefined` in `@react-native/metro-config`, but `metro-config`'s `loadConfig` expects this to be iterable: https://github.com/facebook/metro/blob/main/packages/metro-config/src/loadConfig.js#L326 🤔 

The only way I can get it working is by deleting the `watchFolders` property on the object returned from `getDefaultValues` here: https://github.com/facebook/react-native/blob/f1a7f08feb2a63926e91d37b5d85130c7b74ee31/packages/metro-config/index.js#L92 while omitting the `watchFolders` from the `config` in `@react-native/metro-config` entirely.

I've pushed another branch to illustrate the changes we'll need: https://github.com/kraenhansen/react-native/compare/kh/metro-config-workspace-root...kraenhansen:react-native:kh/metro-config-workspace-root-defaults?expand=1

Reply:
> metro-config's loadConfig expects this to be iterable: [facebook/metro@main/packages/metro-config/src/loadConfig.js#L326](https://github.com/facebook/metro/blob/main/packages/metro-config/src/loadConfig.js?rgh-link-date=2024-01-02T16%3A12%3A59Z#L326)

Interesting! I think we should fix that in Metro as `watchFolders` can be `undefined` as indicated in `InputConfigT`.

---

Separately, and back to the problem at hand (forgive me if I've lost context on something obvious) — can we instead:

- Keep the default of `watchFolders = []`.
- Apply the heuristic if `watchFolders` is `[]` or `== null`, and do nothing otherwise.
    - What's relevant to the user is _have they set an override for `watchFolders` in their `metro.config.js`?_ (i.e. in the [template](https://github.com/facebook/react-native/blob/main/packages/react-native/template/metro.config.js#L9)). If not, then we can apply this :)
    - This would be [documented](https://reactnative.dev/docs/metro) as ~"**`watchFolders` behaviour and workspaces detection**: The default for `watchFolders` is `[]`. If you configure a value to an array containing anything other than this, workspaces detection will not be applied."

Reply:
Okay. I misunderstood you then. I thought you wanted end-users to be able to use an empty array to indicate that they want to opt out of the workspace detection.

The [metro-config's loadConfig injects the project root as a watch folder](https://github.com/facebook/metro/blob/main/packages%2Fmetro-config%2Fsrc%2FloadConfig.js#L325), so from the cli plugin, it won't ever be empty, but instead I suggested to check if it contains any other value than the project root.

Reply:
Hmm - a user may explicitly specify `watchFolders: []` and in that case we should *not* override it by applying auto-detection. A referential check *could* work (to differentiate explicit `[]` from the default `[]`), though the clearest API.

(IMO, this would be easier if `metro.config.js` exported a partial `InputConfigT`, as it used to, rather than a full `ConfigT` with defaults already merged and explicit intent lost, as it does post `@react-native/metro-config`. I don't fully recall why we did it that way @huntie? I think it was discussed and I'm just forgetting the trade-offs.)

Reply:
Hmm. That default config merge is applied inside `@react-native/metro-config`, so we have control. The partial config was our initial approach, pivoted away from here:

- https://github.com/facebook/react-native/pull/38092

Now that we've [documented Metro config in RN better](https://reactnative.dev/docs/metro#advanced-using-a-config-function), perhaps we could move forward with the following changes in `@react-native/metro-config`:
- Removing the default config merge.
- Removing `watchFolders: []` in these defaults.
- Artificially reading and assigning the defaults for `sourceExts` and `assetExts` (to fit user expectations as this is frequently extended from).

Reply:
I finally found some time to get back to this:
- I've implemented a couple of tests, including a failing test for the case where a developer explicitly set their `watchFolder: []` in the `metro.config.js`.
- I've realised, that even if we fix the bug with `metro-config` not taking `watchFolders: undefined`, I would have to move out the injection of `configWithArgs.projectRoot` over to `community-cli-plugin` (to determine if the developer left out the value or explicitly set it to the empty array) .. but that could break someone relying on `metro-config` injecting this 🤔 Any ideas on how to get around that? Or if that would be okay?
I can push a PR to metro to illustrate the change, but basically I would move [the entire `overriddenConfig` part](https://github.com/facebook/metro/blob/18210320c0466499f7874abc26eac1fe60ec519c/packages/metro-config/src/loadConfig.js#L322-L334) into [the `community-cli-plugin`'s overrides](https://github.com/facebook/react-native/blob/07e8ae42bed71f54bbe0e786ccad88b2f14648a4/packages/community-cli-plugin/src/utils/loadMetroConfig.js#L32C10-L32C27).